### PR TITLE
Npm: Map CLI warning / error messages to ORT issues

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

**Review by MarcelBochtler — COMMENTED** [review]
LGTM, but I want to test it internally, as this might create many additional issues.

**Comment by sschuberth** [comment]
> LGTM, but I want to test it internally, as this might create many additional issues.

I've updated the PR with another commit, in case you didn't start testing yet.

**Comment by sschuberth** [comment]
This is now somewhat blocked by https://github.com/oss-review-toolkit/ort/pull/6244 as I need more flexible ways to compare to expected results be leveraging Kotest's data class matchers.

**Review by MarcelBochtler — DISMISSED** [review]

File: analyzer/src/funTest/assets/projects/external/git-repo-expected-output.yml [hunk 1]
```diff
@@ -4311,6 +4311,22 @@ analyzer:
           path: "types/node"
       curations: []
     issues:
+      NPM::isarray:2.0.5:
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/assets/projects/external/git-repo-expected-output.yml:4314]: Commit message: You mention `getPackageReferenceForMissingModule()` but you removed that function two commits earlier?

File: analyzer/src/main/kotlin/managers/Npm.kt [hunk 2]
```diff
@@ -571,15 +571,36 @@ open class Npm(
...
+                issues += OrtIssue(
+                    source = managerName,
+                    message = issueLines.joinToString("\n"),
+                    severity = severity
+                )
+            }
+        }
+
+        mapLinesToOrtIssues("npm WARN ", Severity.WARNING)
+        mapLinesToOrtIssues("npm ERR! ", Severity.ERROR)
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Npm.kt:601]: I would like to have a test case which also tests the error path, I would be fine with adding that later.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: analyzer/src/funTest/assets/projects/external/git-repo-expected-output.yml (hunk 1 above)
Comment (sschuberth) [reply analyzer/src/funTest/assets/projects/external/git-repo-expected-output.yml:4314]: Sorry, I had reordered comments meanwhile. I'll fix the commit message to not refer to that function anymore.

File: analyzer/src/main/kotlin/managers/Npm.kt (hunk 2 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Npm.kt:601]: This case seems to be a bit difficult to trigger consistently across different NPM versions, so I'd indeed prefer to add this later.

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Npm.kt:615]: I found a way to trigger that error and added a test.

Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/Npm.kt:615]: Did you add that already? In the expected results diffs I only see warnings.

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Npm.kt:615]: > Did you add that already?

Yes.

> In the expected results diffs I only see warnings.

I didn't bother to create / modify an external result file, but added a new test that tests this programmatically:

https://github.com/oss-review-toolkit/ort/pull/6243/files#diff-27797998f83fc67dfd200ad5f84575223bca6e55450685f75ca392c3dd0f9753R111-R123

Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/Npm.kt:615]: Thanks, I missed that.

**Review by mnonnenmacher — APPROVED** [review]

**Pull request merged**